File: Holang.Core/Parsing/HolowareParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using Holang.Core.Runtime;

namespace Holang.Core.Parsing;

public sealed class HolowareParser {
    private string _code;
    private int _pos;
    private readonly Holoware _ware = new();
    private string? _ego;
    private readonly bool _startWithSystem;

    public HolowareParser(string code, string? ego = null, bool startWithSystem = false) {
        _code = code;
        _pos = 0;
        _ego = ego;
        _startWithSystem = startWithSystem;
    }

    public Holoware Parse() {
        _code = FilterComments(_code);
        while (_pos < _code.Length) {
            var next = FindNextSpanStart();
            if (next == -1) { ParseText(_code[_pos..]); break; }
            if (next > _pos) ParseText(_code[_pos..next]);
            _pos = next;
            var spanText = ReadUntilSpanEnd();
            ParseSpan(spanText);
        }
        if (_startWithSystem) AddImplicitEgoIfNeeded();
        FinalizeTextSpans();
        return _ware;
    }

    private void AddImplicitEgoIfNeeded() {
        if (!string.IsNullOrEmpty(_ego)) return;
        foreach (var s in _ware.Spans) {
            if (s is TextSpan t) { if (!string.IsNullOrWhiteSpace(t.Text)) { _ware.Spans.Insert(0, new EgoSpan { Ego = "system" }); _ego = "system"; } break; }
            if (s is EgoSpan || s is ContextResetSpan) return;
        }
    }

    private void AddSpan(Span span) {
        var last = _ware.Spans.Count > 0 ? _ware.Spans[^1] : null;
        var isText = span is TextSpan;

        if (string.IsNullOrEmpty(_ego) && span is (SampleSpan or ObjSpan or ClassSpan)) {
            AddImplicitEgoIfNeeded();
            if (string.IsNullOrEmpty(_ego)) throw new InvalidOperationException($"Cannot have {span.GetType().Name} before an ego.");
        }

        if (span is EgoSpan eg) {
            if (_ego == eg.Ego) return; // skip duplicate consecutive ego
            _ego = eg.Ego;
        } else if (span is ContextResetSpan) {
            _ego = null;
        }

        if (isText && last is TextSpan lt) { lt.Text += ((TextSpan)span).Text; return; }
        if (isText && last is not TextSpan && last is not null) ((TextSpan)span).Text = ((TextSpan)span).Text.TrimStart();
        if (isText && string.IsNullOrWhiteSpace(((TextSpan)span).Text)) return;
        _ware.Spans.Add(span);
    }

    private void ParseSpan(string spantext) {
        var buf = new List<Span>();
        BuildSpan(buf, spantext);
        foreach (var s in buf) AddSpan(s);

        // If the last span is a ClassSpan, try to parse an indented body block
        var last = _ware.Spans.Count > 0 ? _ware.Spans[^1] : null;
        if (last is ClassSpan cs) {
            var (body, newPos) = ParseIndentedBlock(_code, _pos);
            if (body is not null) { cs.Body = body; _pos = newPos; }
        }
    }

    private void ParseText(string text) {
        if (string.IsNullOrEmpty(text)) return;
        var processed = text.Replace("\\\\", "\\").Replace("\\<|", "<|");
        if (string.IsNullOrEmpty(processed)) return;
        AddSpan(new TextSpan { Text = processed });
    }

    private int FindNextSpanStart() {
        var pos = _pos;
        while (true) {
            var found = _code.IndexOf("<|", pos, StringComparison.Ordinal);
            if (found == -1) return -1;
            var numBackslashes = 0;
            var i = found - 1;
            while (i >= 0 && _code[i] == '\\') { numBackslashes++; i--; }
            if (numBackslashes % 2 == 1) { pos = found + 1; continue; } // escaped
            return found;
        }
    }

    private (Holoware? body, int newPos) ParseIndentedBlock(string code, int startPos) {
        var (raw, endPos) = ReadIndentedBlockContent(code, startPos);
        if (raw is null) return (null, startPos);
        var dedented = Dedent(raw);
        if (string.IsNullOrWhiteSpace(dedented)) return (null, endPos);
        var parser = new HolowareParser(dedented);
        var body = parser.Parse();
        return (body, endPos);
    }

    private static string Dedent(string block) {
        var lines = block.Replace("\r\n", "\n").Replace("\r", "\n").Split('\n');
        var indents = lines.Where(l => l.Length > 0).Select(l => l.TakeWhile(c => c == ' ').Count()).DefaultIfEmpty(0).Min();
        var sb = new StringBuilder();
        foreach (var line in lines) sb.AppendLine(line.Length >= indents ? line[indents..] : line);
        return sb.ToString();
    }

    private (string? content, int newPos) ReadIndentedBlockContent(string code, int startPos) {
        var lines = code[startPos..].Split(new[] { "\n" }, StringSplitOptions.None);
        if (lines.Length == 0) return (null, startPos);
        var first = lines[0];
        var curPos = startPos;
        if (string.IsNullOrWhiteSpace(first)) { curPos += first.Length + 1; lines = lines.Skip(1).ToArray(); if (lines.Length == 0) return (null, curPos); first = lines[0]; }
        var indentation = first.TakeWhile(c => c == ' ').Count();
        if (indentation == 0) return (null, startPos);
        var blockLines = new List<string>();
        foreach (var line in lines) {
            var indent = line.TakeWhile(c => c == ' ').Count();
            if (string.IsNullOrWhiteSpace(line)) { blockLines.Add(line); curPos += line.Length + 1; continue; }
            if (indent >= indentation) { blockLines.Add(line); curPos += line.Length + 1; }
            else break;
        }
        if (blockLines.Count == 0) return (null, startPos);
        return (string.Join("\n", blockLines), curPos);
    }

    private string ReadUntilSpanEnd() {
        if (_code.AsSpan(_pos).StartsWith("<|")) {
            var start = _pos + 2;
            var end = _code.IndexOf("|>", start, StringComparison.Ordinal);
            if (end == -1) throw new InvalidOperationException("Unclosed tag");
            _pos = end + 2;
            return _code[start..end];
        }
        throw new InvalidOperationException("Not at start of span");
    }

    private void FinalizeTextSpans() {
        Span? last = null;
        for (int i = 0; i < _ware.Spans.Count; i++) {
            var s = _ware.Spans[i];
            if (s is TextSpan t) {
                if (last is TextSpan lt) { lt.Text += t.Text; _ware.Spans.RemoveAt(i); i--; }
                else last = t;
            } else last = s;
        }
    }

    // Grammar / builders
    public static string FilterComments(string content) {
        var lines = content.Replace("\r\n", "\n").Split('\n');
        return string.Join("\n", lines.Where(l => !l.TrimStart().StartsWith('#')));
    }

    private static bool IsEgoOrSampler(string @base, List<string> kargs, Dictionary<string, string> kwargs)
        => @base is "o_o" or "@_@" or "x_x" || kwargs.ContainsKey("fence") || kwargs.ContainsKey("<>");

    private static bool IsContextReset(string @base, List<string> kargs, Dictionary<string, string> kwargs)
        => new[] { "+++", "===", "---", "^^^", "###", "@@@", "\"\"\"", "***", "%%%" }.Contains(@base);

    private static readonly Dictionary<string, string> EgoMap = new() { ["o_o"] = "user", ["@_@"] = "assistant", ["x_x"] = "system" };

    private static (string Base, List<string> KArgs, Dictionary<string, string> KwArgs) ParseSpanTag(string tag) {
        var kargs = new List<string>();
        var kwargs = new Dictionary<string, string>();
        // simple shell-like split covering quotes
        var parts = Regex.Matches(tag, "\"[^\"]*\"|'[^']*'|\\S+").Select(m => m.Value.Trim()).ToList();
        if (parts.Count == 0) return (string.Empty, kargs, kwargs);
        var baseSpan = parts[0];
        foreach (var part in parts.Skip(1)) {
            if (part.Contains('=')) {
                var idx = part.IndexOf('=');
                var key = part[..idx];
                var val = part[(idx + 1)..].Trim('"', '\'');
                kwargs[key] = val;
            } else if (part.StartsWith("<>")) {
                if (part.Length > 2) kwargs["<>"] = part[2..];
                else throw new ArgumentException("Empty <> attribute");
            } else kargs.Add(part);
        }
        return (baseSpan, kargs, kwargs);
    }

    private static void BuildSpan(List<Span> outList, string spantext) {
        var tag = spantext.Trim();
        if (tag.Length == 0) return;
        var (baseSpan, kargs, kwargs) = ParseSpanTag(tag);

        if (IsEgoOrSampler(baseSpan, kargs, kwargs)) { BuildEgoOrSampler(outList, baseSpan, kargs, kwargs); return; }
        if (IsContextReset(baseSpan, kargs, kwargs)) { BuildContext(outList, train: baseSpan == "+++"); return; }

        if (!string.IsNullOrEmpty(baseSpan)) {
            if (char.IsUpper(baseSpan[0])) BuildClass(outList, baseSpan, kargs, kwargs);
            else BuildObj(outList, baseSpan, kargs, kwargs);
        }
    }

    private static void BuildClass(List<Span> outList, string @base, List<string> kargs, Dictionary<string, string> kwargs) {
        var span = new ClassSpan { ClassName = @base };
        span.WithArgs(kargs, kwargs, new());
        span.Body = null;
        outList.Add(span);
    }

    private static void BuildEgoOrSampler(List<Span> outList, string @base, List<string> kargs, Dictionary<string, string> kwargs) {
        var parts = @base.Split(':', 2);
        var ego = parts[0];
        var spanId = parts.Length > 1 ? parts[1] : string.Empty;
        outList.Add(new EgoSpan { Ego = EgoMap.TryGetValue(ego, out var mapped) ? mapped : ego, Uuid = spanId });

        var samplerKw = kwargs.Where(kv => kv.Key != "<>" && kv.Key != "fence").ToDictionary(kv => kv.Key, kv => kv.Value);
        var fence = kwargs.GetValueOrDefault("<>") ?? kwargs.GetValueOrDefault("fence") ?? string.Empty;
        if (kargs.Count > 0 || samplerKw.Count > 0 || !string.IsNullOrEmpty(fence)) {
            if (!string.IsNullOrEmpty(fence)) {
                outList.Add(new SampleSpan { Uuid = spanId, Id = spanId, KArgs = new List<string>(kargs), KwArgs = samplerKw, Fence = fence });
            }
        }
    }

    private static void BuildContext(List<Span> outList, bool train) { outList.Add(new ContextResetSpan(train)); }

    private static void BuildObj(List<Span> outList, string @base, List<string> kargs, Dictionary<string, string> kwargs) {
        var ids = @base.Split('|').Select(v => v.Trim()).ToList();
        var span = new ObjSpan { VarIds = ids };
        span.WithArgs(kargs, kwargs, new());
        outList.Add(span);
    }
}

File: Holang.Core/holoware/holoware_parser.py
import shlex
import textwrap
from typing import Dict, Optional, Tuple

from rich.panel import Panel

from errloom.lib.log import indent_decorator
from errloom.lib import log
from errloom.holoware.holoware import (
    ClassSpan,
    ContextResetSpan,
    EgoSpan,
    Holoware,
    ObjSpan,
    SampleSpan,
    Span,
    TextSpan,
)

logger = log.getLogger(__name__)

# --- Grammar Definition ---
def is_ego_or_sampler(base, kargs, kwargs) -> bool:
    return base in ("o_o", "@_@", "x_x") or "fence" in kwargs or "<>" in kwargs

def is_context_reset(base, kargs, kwargs) -> bool:
    return base in ("+++", "===", "---", "^^^", "###", "@@@", "\"\"\"", "***", "%%%")

HOLOWARE_GRAMMAR: list[Dict] = [
    {"match": is_ego_or_sampler, "handler": lambda *args: _build_ego_or_sampler(*args)},
    {"match": is_context_reset, "handler": lambda *args: _build_context(train=args[1] == "+++")(*args)},
]

EGO_MAP = {"o_o": "user", "@_@": "assistant", "x_x": "system"}

class HolowareParser:
    def __init__(self, code: str, ego=None, start_with_system=False):
        self.code = code
        self.pos = 0
        self.ware = Holoware()
        self.ego: Optional[str] = ego
        self.start_with_system = start_with_system

    def parse(self) -> "Holoware":
        logger.push_debug("PARSE")
        self.code = filter_comments(self.code)

        # if self.code.strip() and not self.code.lstrip().startswith('<|'):
        #     self.ego = 'system'
        #     self.ware.spans.append(EgoSpan(ego='system'))

        while self.pos < len(self.code):
            next_span_pos = self._find_next_span_start()

            if next_span_pos == -1:
                self._parse_text(self.code[self.pos:])
                break

            if next_span_pos > self.pos:
                self._parse_text(self.code[self.pos:next_span_pos])

            self.pos = next_span_pos
            spantext = self.read_until_span_end()
            self._parse_span(spantext)

        if self.start_with_system:
            self._add_implicit_ego_if_needed()

        self._finalize_text_span()

        logger.pop()
        return self.ware

    def _add_implicit_ego_if_needed(self):
        # Implicitly create a system ego if there's text content before any ego is set.
        if self.ego:
            return

        text_span_found = False
        for s in self.ware.spans:
            if isinstance(s, TextSpan):
                if s.text.strip():
                    text_span_found = True
                break
            if isinstance(s, (EgoSpan, ContextResetSpan)):
                # Don't add implicit ego if an explicit one is already there
                return

        if text_span_found:
            new_ego = EgoSpan(ego='system')
            self.ware.spans.insert(0, new_ego)
            self.ego = 'system'

    def _add_span(self, span: Span):
        """Adds a span to the current holoware object."""
        logger.debug(f"[bold green]{type(span).__name__}[/]")
        last_span = self.ware.spans[-1] if self.ware.spans else None
        is_text = isinstance(span, TextSpan)

        # requires_ego_insertion = not self.ego and is_text and span.text.strip()
        # if self.start_with_system and requires_ego_insertion:
        #     new_ego = EgoSpan(ego='system')
        #     self.ware.spans.append(new_ego)
        #     self.ego = new_ego.ego
        #     last_span = new_ego

        if not self.ego and isinstance(span, (SampleSpan, ObjSpan, ClassSpan)):
            self._add_implicit_ego_if_needed()
            if not self.ego:
                raise ValueError(f"Cannot have {type(span).__name__} before a ego.")

        if isinstance(span, EgoSpan):
            # Avoid creating duplicate, consecutive ego spans.
            if self.ego == span.ego:
                return
            self.ego = span.ego
        elif isinstance(span, ContextResetSpan):
            self.ego = None  # Context resets clear the current ego.

        # Merge consecutive text spans for cleaner output.
        if is_text and isinstance(last_span, TextSpan):
            last_span.text += span.text
            return

        # Remove leading whitespace from text that follows any span to keep it clean.
        if is_text and not isinstance(last_span, TextSpan) and last_span is not None:
            span.text = span.text.lstrip()

        # Don't add text spans that are only whitespace
        if is_text and not span.text.strip():
            return

        self.ware.spans.append(span)

    @indent_decorator("SPAN", log_func=logger.debug)
    def _parse_span(self, spantext: str):
        # log.push("PARSE", f"<|{spantext}|>")
        spanbuf = []
        build_span(spanbuf, spantext)
        for span in spanbuf:
            self._add_span(span)

        # check for an indented block to be its body.
        last_span = self.ware.spans[-1] if self.ware.spans else None
        if isinstance(last_span, ClassSpan):
            body_holoware, new_pos = self._parse_indented_block(self.code, self.pos)
            if body_holoware:
                last_span.body = body_holoware
                self.pos = new_pos
        # log.pop()

    @indent_decorator("TEXT", log_func=logger.debug)
    def _parse_text(self, text: str):
        if not text:
            return

        # Unescape backslashes and tags
        processed_text = text.replace('\\\\', '\\').replace('\\<|', '<|')

        if not processed_text:
            return

        logger.debug(f"{repr(processed_text[:40])}")
        span = TextSpan(text=processed_text)
        self._add_span(span)

    def _find_next_span_start(self) -> int:
        pos = self.pos
        while True:
            found_pos = self.code.find("<|", pos)
            if found_pos == -1:
                logger.debug(f"no more spans from pos {self.pos}")
                return -1

            num_backslashes = 0
            i = found_pos - 1
            while i >= 0 and self.code[i] == '\\':
                num_backslashes += 1
                i -= 1

            if num_backslashes % 2 == 1:
                # Escaped, continue searching
                logger.debug(f"found escaped span at {found_pos}, continuing search from {found_pos + 1}")
                pos = found_pos + 1
                continue

            logger.debug(f"found span at {found_pos}")
            return found_pos

    @indent_decorator("BLOCK", log_func=logger.debug)
    def _parse_indented_block(self, code: str, start_pos: int) -> Tuple[Optional[Holoware], int]:
        block_content, end_pos = self._read_indented_block_content(code, start_pos)
        if block_content is None:
            logger.debug("[block] x (no content)")
            # log.pop()
            return None, start_pos

        # --- Dedent and Prepare for Parsing ---
        dedented_block = textwrap.dedent(block_content)

        logger.debug(Panel(
            dedented_block.strip(),
            title="Parsing Indented Block",
            border_style="dim",
            expand=False,
            padding=(1, 4)
        ))
        # Check if the block is empty or contains only whitespace
        if not dedented_block.strip():
            logger.debug("x (empty after dedent)")
            # log.pop()
            return None, end_pos

        # --- Recursive Parsing ---
        # Start a fresh parser for the block without inheriting the outer ego.
        # This allows the nested body to add an implicit system ego when appropriate.
        parser = HolowareParser(dedented_block)
        body = parser.parse()
        return body, end_pos

    def _read_indented_block_content(self, code: str, start_pos: int) -> Tuple[Optional[str], int]:
        """Reads an indented block of text, returning the content and new position."""
        lines = code[start_pos:].splitlines(True)
        if not lines:
            return None, start_pos

        first_line = lines[0]
        if not first_line.strip():  # Skip empty line after span
            start_pos += len(first_line)
            lines.pop(0)
            if not lines:
                return None, start_pos
            first_line = lines[0]

        indentation = len(first_line) - len(first_line.lstrip(' '))
        if indentation == 0:
            return None, start_pos

        block_lines = []
        current_pos = start_pos
        for line in lines:
            line_indent = len(line) - len(line.lstrip(' '))
            if line.strip() == "":  # allow empty lines within block
                block_lines.append(line)
                current_pos += len(line)
                continue
            if line_indent >= indentation:
                block_lines.append(line)
                current_pos += len(line)
            else:
                break

        if not block_lines:
            return None, start_pos

        return "".join(block_lines), current_pos

    def read_until_span_end(self) -> str:
        if self.code[self.pos:self.pos + 2] != '<|':
            raise ValueError("Not at the start of a span")

        start = self.pos + 2
        try:
            end = self.code.index("|>", start)
            self.pos = end + 2
            return self.code[start:end]
        except ValueError:
            raise ValueError("Unclosed tag")

    def _finalize_text_span(self):
        """Merges consecutive TextSpans and removes leading whitespace."""
        last_span = None
        for i, span in enumerate(self.ware.spans):
            if isinstance(span, TextSpan):
                if last_span and isinstance(last_span, TextSpan):
                    last_span.text += span.text
                    self.ware.spans.pop(i)  # Remove the duplicate
                else:
                    last_span = span
            else:
                last_span = span

# --- Span Builders ---

def parse_span_tag(tag: str) -> Tuple[str, list[str], Dict[str, str]]:
    kwargs: Dict[str, str] = {}
    kargs: list[str] = []

    parts = shlex.split(tag)
    if not parts:
        return "", [], {}

    base_span = parts[0]
    new_kargs = []
    for part in parts[1:]:
        if "=" in part:
            key, value = part.split("=", 1)
            kwargs[key] = value
        elif part.startswith('<>'):
            if len(part) > 2:
                kwargs['<>'] = part[2:]
            else:
                raise ValueError("Empty <> attribute")
        else:
            new_kargs.append(part)

    kargs.extend(new_kargs)

    return base_span, kargs, kwargs

def build_span(out: list[Span], spantext: str):
    """Finds the correct handler in the grammar and creates span(s) for a tag."""
    tag_content = spantext.strip()
    if not tag_content:
        return
    base, kargs, kwargs = parse_span_tag(tag_content)

    logger.debug(f"base='{base}' kargs={kargs} kwargs={kwargs}")

    for rule in HOLOWARE_GRAMMAR:
        if rule['match'](base, kargs, kwargs):
            logger.debug(f"handler={rule['match'].__name__}")
            rule['handler'](out, base, kargs, kwargs)
            return

    # Fallback for ObjSpans or unhandled ClassSpans
    if base:
        if base[0].isupper():
            logger.debug("handler=_build_class_span ClassSpan")
            _build_class(out, base, kargs, kwargs)
        else:
            logger.debug("handler=_build_class_span ObjSpan")
            _build_obj(out, base, kargs, kwargs)


def _build_class(out: list[Span], base, kargs, kwargs):
    """Handler for creating ClassSpans."""
    span = ClassSpan(class_name=base)
    span.set_args(kargs, kwargs, {})
    span.body = None
    out.append(span)

def _build_ego_or_sampler(out: list, base: str, kargs: list, kwargs: dict):
    # Split ego and potential identifier after colon
    parts = base.split(":", 1)
    ego = parts[0]
    span_id = parts[1] if len(parts) > 1 else ""

    out.append(EgoSpan(ego=EGO_MAP.get(ego, ego), uuid=span_id))

    # A sampler can be defined with kargs/kwargs on the ego span
    sampler_kwargs = {k: v for k, v in kwargs.items() if k not in ("<>", "fence")}
    fence = kwargs.get("<>", "") or kwargs.get("fence", "")
    if kargs or sampler_kwargs or fence:
        if not fence:
            # It's not a sampler if it doesn't have a fence
            # This can happen if there are only kargs, which are not supported for samplers.
            # We just ignore them.
            pass
        else:
            # Assign the human-readable id (span_id) universally onto SampleSpan.id
            out.append(SampleSpan(uuid=span_id, id=span_id, kargs=kargs, kwargs=sampler_kwargs, fence=fence))

def _build_context(train: bool):
    def _handler(out: list, base: str, kargs: list, kwargs: dict):
        """Handler for creating ContextResetSpans."""
        out.append(ContextResetSpan(train=train))

    return _handler

def _build_obj(out: list[Span], base, kargs, kwargs):
    """Fallback handler for creating ObjSpans from one or more IDs."""
    logger.debug("")
    var_ids = [v.strip() for v in base.split('|')]
    span = ObjSpan(var_ids=var_ids).set_args(kargs, kwargs, {})
    out.append(span)


def filter_comments(content: str) -> str:
    """
    Removes comments from holoware content.
    Only supports full-line comment starting with #
    """
    lines = content.split('\n')
    processed_lines = []
    for line in lines:
        if not line.strip().startswith('#'):
            processed_lines.append(line)
    return "\n".join(processed_lines)

